### PR TITLE
Implement Code Formatting

### PR DESCRIPTION
**Pull request by kstatz12**
This is my first stab at implementing code formatting features from roslyn. I tried to be as true to what the omnisharp team did. the thing that is missing is the OmnisharpComment formatting bits because i couldnt find the source for it

Cheers

### CODE REVIEW

**Comment by kstatz12** [comment]
Side note: how do you test this thing.

**Comment by razzmatazz** [comment]
Great! Will take a look.

by testing you mean unit-testing or testing in an editor

I have ideas how to do unit testing but underlying ionide.lsp lsp lib is still in flux while thinking how to sequence rw vs r/o rpc calls and whether to remake the code into a more functional way.. see my latest commits wrt to sync issues and incremental sync 

In other words unit testing should wait a bit until lsp bits are more stable

File: src/CSharpLanguageServer/Server.fs [hunk 1]
```diff
@@ -852,6 +856,36 @@ type CSharpLspServer(lspClient: CSharpLspClient, options: Options) =
...
+             return formattingChanges |> Some |> success
+        }
+
+    override __.TextDocumentOnTypeFormatting(format: DocumentOnTypeFormattingParams) : AsyncLspResult<TextEdit [] option> =
+        async {
+            let maybeDocument = getDocumentForUri state format.TextDocument.Uri
+            let! formattingChanges =
+                match maybeDocument with
+                | Some doc -> handleTextOnTypeFormatAsync doc format.Ch format.Position
+                | None -> async { return Array.empty<Types.TextEdit> }
```
Comment (razzmatazz) [inline src/CSharpLanguageServer/Server.fs:885]: I think you can just do
```
| None -> async { return [] }
```

or maybe there is an even better way (but I am newbie in F#).. maybe there is some helper to do this
```
| None -> [] |> Async.Result
```

File: src/CSharpLanguageServer/Server.fs [hunk 2]
```diff
@@ -852,6 +856,36 @@ type CSharpLspServer(lspClient: CSharpLspClient, options: Options) =
 
         async { return metadataMaybe |> success }
 
+    override __.TextDocumentFormatting(format: Types.DocumentFormattingParams) : AsyncLspResult<Types.TextEdit [] option> =
```
Comment (razzmatazz) [inline src/CSharpLanguageServer/Server.fs:859]: I rewrote request handling a bit to sequence r/w vs r/o rpc calls and you need to source `state` with `withStateRead` -- that is way it does not compile with the latest code

File: src/CSharpLanguageServer/Server.fs (hunk 1 above)
Comment (razzmatazz) [reply src/CSharpLanguageServer/Server.fs:885]: same thing on other places -- no need to to `Array.empty<>` -- F# uses type inference more than C# and you can just type `[]` in those places

Comment (kstatz12) [reply src/CSharpLanguageServer/Server.fs:949]: So thats what I thought too but the compiler throws `error FS0001: This expression was expected to have type    'TextEdit[]'    but here has type    ''a list` with the 
```
| None -> async { return [] }
```
syntax

File: src/CSharpLanguageServer/Server.fs (hunk 2 above)
Comment (kstatz12) [reply src/CSharpLanguageServer/Server.fs]: Sounds good, ill fetch from upstream and make that change

**Comment by razzmatazz** [comment]
Works great for me! Thanks!

Will release later when I have state management fixed as my latest fixes for incremental buffer update don't work properly all the time due to me using asyncrwlock which does not actually sequence operations properly..

Merging

**Comment by kstatz12** [comment]
That was fun, reach out if there are any other backlog items you wouldn't mind me working on

**Comment by razzmatazz** [comment]
oh, feel free to take anything you feel you want to tinker with -- I have my own wishlist here:
- https://github.com/razzmatazz/csharp-language-server#todo-list

and you may have your own itches to scratch which is not on the list.. so..

**Comment by razzmatazz** [comment]
Hey @kstatz12 I just released 0.2.0 with your changes

This releases contains quite a large amount of changes so hopefully we didn't break much :E

**Pull request merged**